### ISSUE TITLE
On Windows 10 (current version) you cannot close the socket and continue to read data

### ISSUE THREAD

**Issue by RichardCroxall** [enwi/hueplusplus#51]
I am writing a home automation system for a Raspberry Pi (linux) but using a PC with Visual Studio as my development system.
I experience a runtime error using the following block of code in WinHttpHandler.cpp at line 144:

    // shutdown the connection for sending since no more data will be sent
    // the client can still use the ConnectSocket for receiving data
    if (shutdown(connect_socket, SD_SEND) == SOCKET_ERROR)
    {
        int err = WSAGetLastError();
        std::cerr << "WinHttpHandler: shutdown failed: " << err << std::endl;
        throw(std::system_error(err, std::system_category(), "WinHttpHandler: shutdown failed"));
    }

It transpires you can not (or no longer) do this with Windows 10.
I moved the block else where as shown in the attached file.
I also added a couple of #defines to cut down on the ridiculously large number of warning messages.

#define _WINSOCK_DEPRECATED_NO_WARNINGS 1
#define _CRT_SECURE_NO_WARNINGS 1

I have pasted the block of code as shown:

    // Send an initial buffer
    if (::send(connect_socket, msg.c_str(), msg.size(), 0) == SOCKET_ERROR)
    {
        int err = WSAGetLastError();
        std::cerr << "WinHttpHandler: send failed: " << err << std::endl;
        throw(std::system_error(err, std::system_category(), "WinHttpHandler: send failed"));
    }

    // shutdown the connection for sending since no more data will be sent
    // the client can still use the ConnectSocket for receiving data
    if (shutdown(connect_socket, SD_SEND) == SOCKET_ERROR)
    {
        int err = WSAGetLastError();
        std::cerr << "WinHttpHandler: shutdown failed: " << err << std::endl;
        throw(std::system_error(err, std::system_category(), "WinHttpHandler: shutdown failed"));
    }

![winhttphandler](https://user-images.githubusercontent.com/16840482/85849007-13068200-b7a2-11ea-9a03-6c4910e764fb.JPG)

**Comment by Jojo-1000** [comment]
That is weird. What kind of error are you getting? In [the docs](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-shutdown#remarks) exactly this order is mentioned for a graceful disconnect. I also believe I successfully tested this under Win 10 at some point.

**Comment by RichardCroxall** [comment]
That was several months ago and I did not keep a record of the error
message, but I remember the error description was fairly specific about the
socket being closed and the VS2019 debugger pinpointed the offending lines
of code.
I am sure this worked once on Windows 10, but everytime they have a major
refresh, things stop working.
To my mind it would always be something you would expect to break sometime
under some implementation, so to me it seems a non-brainer to move the
lines to where the socket is still open.
The change was very easy and once I worked out what was going on, it only
took me a couple of minutes.

**Comment by Jojo-1000** [comment]
> The shutdown function does not close the socket. Any resources attached to the socket will not be freed until closesocket is invoked.

> To assure that all data is sent and received on a connected socket before it is closed, an application should use shutdown to close connection before calling closesocket

This is why I would leave it as it is right now. I see no reason why it would not work, and the problem is that the data may not be received before the socket is closed. I don't think it is likely that windows introduces a breaking change in a decade old API without leaving a note on the docs pages.

**Comment by RichardCroxall** [comment]
Hi Jan, I have redeveloped my application using your development branch (the version available on 26/6/2020 about 12:00).
It appears I hadn't remembered this issue quite correctly. My call to findBridges
    std::vector<hueplusplus::HueFinder::HueIdentification> bridgeHueIdentifications = hue_finder.FindBridges();
raises an exception because there is 0 bytes returned from
        res = recv(connect_socket, recvbuf, recvbuflen, 0);
at line 157 in WinHttpHandler.cpp.
This causes an exception to be raised at line 40 in BaseHttpHandler.cpp
        throw HueException(CURRENT_FILE_INFO, "Failed to find body in response");
which my code catches and then retries a few more times before giving up.

Moving this block to the end of WinHttpHandler::send() just before the "return response;"

    // *** shutdown the connection for sending since no more data will be sent
    // *** the client can still use the ConnectSocket for receiving data
    // Move socket closing code to after we have received data as closing it before stopped data being received RJC 18/04/2020
    if (shutdown(connect_socket, SD_SEND) == SOCKET_ERROR)
    {
        int err = WSAGetLastError();
        std::cerr << "WinHttpHandler: shutdown failed: " << err << std::endl;
        throw(std::system_error(err, std::system_category(), "WinHttpHandler: shutdown failed"));
    }
	
Now response contains:
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 28 Jun 2020 17:21:04 GMT
Content-Type: text/xml
Connection: close
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Content-Security-Policy: default-src 'self'
Cache-Control: no-store
Pragma: no-cache
Referrer-Policy: no-referrer

<?xml version="1.0" encoding="UTF-8" ?>
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
<major>1</major>
<minor>0</minor>
</specVersion>
<URLBase>http://192.168.1.113:80/</URLBase>
<device>
<deviceType>urn:schemas-upnp-org:device:Basic:1</deviceType>
<friendlyName>Philips hue (192.168.1.113)</friendlyName>
<manufacturer>Signify</manufacturer>
<manufacturerURL>http://www.meethue.com</manufacturerURL>
<modelDescription>Philips hue Personal Wireless Lighting</modelDescription>
<modelName>Philips hue bridge 2015</modelName>
<modelNumber>BSB002</modelNumber>
<modelURL>http://www.meethue.com</modelURL>
<serialNumber>001788a649f0</serialNumber>
<UDN>uuid:2f402f80-da50-11e1-9b23-001788a649f0</UDN>
<presentationURL>index.html</presentationURL>
<iconList>
<icon>
<mimetype>image/png</mimetype>
<height>48</height>
<width>48</width>
<depth>24</depth>
<url>hue_logo_0.png</url>
</icon>
</iconList>
</device>
</root>

I am using Microsoft Visual Studio Community 2019 Version 16.6.2. with Microsoft .NET framework version 4.8.0352.
I am running in Windows 10 Home Version 1909 installed on 04/03/2020. OS Build 18363.900.

With the modification suggested above, the code runs fine on Windows 10. 
Of course this isn't used on the Linux code as that uses LinHttpHandler.cpp, so it doesn't affect my Raspberry Pi.
However I don't usually bother to test something on the Pi until it works on Windows/VS2019 as I don't have a debugger on the Pi.

I hope that extra information helps you and gives you more confidence in my suggested fix. You should at least be able to reproduce the problem.

**Comment by Jojo-1000** [comment]
This example does everything in the same order as we do:
[https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recv](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recv)

Does the error occur every single attempt or is it random?
Is it printing no response at all in the error message, or does the response have a header but no body?
Maybe there are other UPnP devices in your network which are also checked, but they somehow block the requests or something.

I believe the error has to be somewhere else. I would not count on the client having received all data without an explicit FIN which is sent by shutdown.

**Comment by RichardCroxall** [comment]
Hi Jan,

The error seemed to occur every time. I tried it about 10 times in the debugger trying to track down the problem before I discovered it was the same problem I had seen before.

 

Each time, the response returned was empty; 0 bytes.

I had two devices within my network which returned an initial response, but the filtering on a string homed in exactly on the correct Philips Hue server.

 

When I move the block of code as indicated, the response seems to be complete and certainly contains the string the code is searching for.

 

I’ve had a quick look at the example and the WinHttpHandler.cpp code and they are difficult to align.

I will compare the code more carefully tomorrow.

 

Good night!

Richard

**Comment by RichardCroxall** [comment]
Hi Jan, I extracted the routine from the example and the routine from the WinHttpHandler and compared.  The code is so similar, it looks like the WinHttpHandler code was originally derived from the example but has received C++ improvements since. However on a line by line comparison, I discovered line 130 -> SocketCloser closeSocket(connect_socket); which appears not to be required (as we don't want to be closing the socket at this point) and is NOT in the reference example from Microsoft. Commenting out this line, allows the code to function again. To be clear, this is the code line just before the "::send" call in std::string WinHttpHandler::send.

So I think removing this line which is not in the reference example solves the problem. (Maybe it was intended as part of the error recovery somewhere else.)
